packages/util/src/logging.ts

```diff
@@ -0,0 +1,36 @@
+import { Logger } from 'ts-log';
+
+export const loggerMethodNames: (keyof Logger)[] = ['debug', 'error', 'fatal', 'info', 'trace', 'warn'];
+
+/**
+ * Appends an object containing context metadata to each log entry
+ * Nested contexts can be added by deriving multiple times.
+ *
+ * Example: single context
+ *   - const oneContextLogger = contextLogger('ctx1', rawLogger);
+ *   - oneContextLogger.debug('hi') calls `rawLogger({context: ['ctx1']}, hi)
+ *
+ * Example: nested context
+ *   - const twoContextLogger = contextLogger('ctx2', oneContextLogger);
+ *   - twoContextLogger.debug('hi') calls `rawLogger({context: ['ctx1', 'ctx2']}, hi)
+ *
+ * Example: nested context with merged message object
+ *   - const mergeLogger = contextLogger('ctx2', oneContextLogger);
+ *   - mergeLogger.debug({info: 'Superb'}, 'hi') calls `rawLogger({context: ['ctx1', 'ctx2'], moreInfo: 'Superb'}, hi)
+ */
+export const contextLogger = (logger: Logger, context: string): Logger => ({
+  context: [...(logger.context || []), context],
+  ...(<Logger>(<unknown>Object.fromEntries(
+    loggerMethodNames.map<[keyof Logger, Logger['info']]>((method) => [
+      method,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (message?: any, ...optionalParams: any[]) => {
+        const contextArg = { context: [...(logger.context || []), context] };
+        if (typeof message === 'object') {
```

Comment:
In case of structured log, this is interesting! But in case of human readable log this will change the line layout depending on the type of the first logged argument making the whole log confused and harder to read.

I would remove this condition and I would only use the `logger[method](contextArg, message, ...optionalParams)` form (the `return` statement is not required here).

Reply:
Same comment here, regarding the special meaning of object type first arg in `bunyan`

Reply:
Didn't thought to that feature... probably because I don't like it... it's just a choice...

Anyway the current implementation replaces the first object with something else loosing both the type and the not own properties of the logged object.
That will prevent the ability to distinguish between an `Error` a `TypeError` or a `ProviderError` and moreover the `message` and `stack` properties.

This should not happen regardless from the choice.

Reply:
Actually it doesn't lose it. It merges the context object with the first arg, when the first arg is an object. But I do agree to not liking it :)

Reply:
> But in case of human readable log this will change the line layout depending on the type of the first logged argument making the whole log confused and harder to read.

Good point - would be most convenient to have context readable in a single line.

> If we add a formatted context string as first arg, we shadow this functionality to users wanting to derive from bunyan

I think it's fine, unless we want to use this bunyan feature internally. If we do, then we can extend `contextLogger` function to support it.

Reply:
Even if the first implementation was really nice as well, I think the #2 version is closer to what we need.
GJ @mirceahasegan !